### ISSUE TITLE
False positive of incompatible types (about generic protocol) since 1.1.337

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#6767]
**Describe the bug**

See sample code below.

In function `f`, `v1` can be assigned to `v0` and `v2` can be assigned to `v1`, but `v2` cannot be directly assigned to `v0`.

**Code or Screenshots**

```python
from typing import Generic, Protocol, TypeVar

V_co = TypeVar("V_co", covariant=True)

class C(Generic[V_co]):
    def f(self, /) -> V_co: ...

class P(Protocol[V_co]):
    def f(self, /) -> V_co: ...

def f(
    v0: C[C[object]],
    v1: C[P[object]],
    v2: P[P[object]],
):
    _ = v0, v1, v2

    if 0:
        v1 = v0
    elif 0:
        v2 = v1
    elif 0:
        v2 = v0  # False report.
```

[Pyright Playground](https://pyright-play.net/?strict=true&locale=en&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSEgMYA0UACuDGBWADY0AqiRAagIYgBYAFDDuAfWZQAvFE4Ie-ABQAicc2U1mAN35JeKGFPYgArkQCUw4RRa8AznagBhRcVLkKAbTVgAuuYAuYSgQqAATImAoYEU7IhZgGgB6cygAWgA%2BKB8AqAA6AqshG3tHWkV6MEZmFm8JP0Dg0IiomLiE5NTM7PrcgryilujFJpCtAAZcp09psAAjACsiChhfXypRqC0ARinPWk95pZW1jaFQrYAmXIODo%2BXV9eFG89CxaS3xmh3vy6KLpBRSabC47D4TTbxQFQYGvC5jS7g7aQljQ2Hw0JaREyCYhADEUAAYrwWHEoCAiDg8AMhEA)

**VS Code extension or command-line**

This issue appears since version 1.1.337. It doesn't appear with version from 1.1.292 through 1.1.336.

**Comment by Azureblade3808** [comment]
If the sample code is modified as below --

```python
from typing import Generic, Protocol, TypeVar

V_co = TypeVar("V_co", covariant=True)

class C(Generic[V_co]):
    def f(self, /) -> V_co: ...

class P0(Protocol[V_co]):
    def f(self, /) -> V_co: ...

class P1(Protocol[V_co]):
    def f(self, /) -> V_co: ...

def f(
    v0: C[C[object]],
    v1: C[P0[object]],
    v2: P1[P0[object]],
):
    _ = v0, v1, v2

    if 0:
        v1 = v0
    elif 0:
        v2 = v1
    elif 0:
        v2 = v0
```

[Pyright Playground](https://pyright-play.net/?strict=true&locale=en&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSEgMYA0UACuDGBWADY0AqiRAagIYgBYAFDDuAfWZQAvFE4Ie-ABQAicc2U1mAN35JeKGFPYgArkQCUw4RRa8AznagBhRcVLkKAbTVgAuuYAuYSgQqAATImAoYEU7IhZgGgB6cygAWgA%2BKB8AqAA6AqshG3tHWgAGRXowRmYWbwk-QODQiKiYuITk1MzsxtyCvKKShzoARiqGJlYG5n8goVDwyOjY%2BMSoFPSsnPzCkSE21ZaQrXLcp09LsAAjACsiChhfXyoTqC0xi88Kz1uHp4vN6LUJaABMuVoYx%2B5T%2B90ez1ewmaIJCYmkH3KNE%2B2LBRSWSCi53eS0%2BGLO73ihKgxNRSw%2BYPJY0pLGptPpoMZMgpQiAA)

No error is reported.

So I suspict there is some data pollution among generic types of a same protocol.

**Comment by erictraut** [comment]
Thanks, I really appreciate the work you put into writing great bug reports.

I've narrowed the regression to [this commit](https://github.com/microsoft/pyright/commit/6758be486575d43ec8a92c5bbae52f6211465b53). The issue is related to the constraint solver, which is conflating the type variable of the "inner" `P` with that of the "outer" `P` in the annotation `P[P[object]]`.

**Comment by Azureblade3808** [comment]
I'm glad to help.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.364.
